src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HandshakeProtocol.cs

```diff
@@ -0,0 +1,136 @@
+
+        private const string ProtocolPropertyName = "protocol";
+        private const string ErrorPropertyName = "error";
+
+        public static void WriteRequestMessage(HandshakeRequestMessage requestMessage, Stream output)
+        {
+            using (var writer = CreateJsonTextWriter(output))
+            {
+                writer.WriteStartObject();
+                writer.WritePropertyName(ProtocolPropertyName);
+                writer.WriteValue(requestMessage.Protocol);
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        public static void WriteResponseMessage(HandshakeResponseMessage responseMessage, Stream output)
+        {
+            using (var writer = CreateJsonTextWriter(output))
+            {
+                writer.WriteStartObject();
+                if (!string.IsNullOrEmpty(responseMessage.Error))
+                {
+                    writer.WritePropertyName(ErrorPropertyName);
+                    writer.WriteValue(responseMessage.Error);
+                }
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        private static JsonTextWriter CreateJsonTextWriter(Stream output)
+        {
+            return new JsonTextWriter(new StreamWriter(output, _utf8NoBom, 1024, leaveOpen: true));
+        }
+
+        public static bool TryParseResponseMessage(ReadOnlySpan<byte> input, out HandshakeResponseMessage responseMessage)
+        {
+            if (!TextMessageParser.TryParseMessage(ref input, out var payload))
+            {
+                throw new InvalidDataException("Unable to parse payload as a handshake message.");
+            }
+
+            using (var memoryStream = new MemoryStream(payload.ToArray()))
+            {
+                using (var reader = new JsonTextReader(new StreamReader(memoryStream)))
+                {
+                    var token = JToken.ReadFrom(reader);
+                    var handshakeJObject = JsonUtils.GetObject(token);
+                    var error = JsonUtils.GetOptionalProperty<string>(handshakeJObject, ErrorPropertyName);
+                    responseMessage = new HandshakeResponseMessage(error);
+                }
+            }
+            return true;
+        }
+
+        public static bool TryParseResponseMessage(ReadOnlySequence<byte> buffer, out HandshakeResponseMessage responseMessage, out SequencePosition consumed, out SequencePosition examined)
+        {
+            if (!TryReadMessageIntoSingleSpan(buffer, out consumed, out examined, out var span))
+            {
+                responseMessage = null;
+                return false;
+            }
+
+            return TryParseResponseMessage(span, out responseMessage);
+        }
+
+        public static bool TryParseRequestMessage(ReadOnlySpan<byte> input, out HandshakeRequestMessage requestMessage)
```

Comment:
Now that we have a `TryReadMessageIntoSingleSpan` method, I think we can actually just inline this into the version that takes `ReadOnlySequence<byte>`. Eventually we'll be building a `Stream` over `ReadOnlySequence` and passing that directly into `JsonTextReader` anyway.